pom.xml

```diff
@@ -175,6 +180,15 @@
         <!-- Inherited from oss-base. Generate PackageVersion.java.-->
         <groupId>com.google.code.maven-replacer-plugin</groupId>
         <artifactId>replacer</artifactId>
+        <executions>
+          <execution>
+            <id>generate-package-version-java</id>
```

Comment:
I wish I understood better how this ties to execution, and whether it would work for Eclipse as well (or if there is even a way to test this without attempting to import from IDE). I understand what this is to fix, I think (trigger replacer upon import, not just `mvn build` from cli), but somehow can not merge back to 2.x in a way that it actually works. Part of the problem is that settings in 2.x (master is for 3.0.0-SNAPSHOT) are divided between `oss-parent` (highest-level, non-jackson-specific parent pom), `jackson-parent` (intermediate) and `jackson-base` (under `jackson-bom` repo). 3.x simplifies this by dropping `jackson-parent`.

I'll try to read more about Maven phase/goal handling for plugins, but any pointers would be appreciated, as this sounds like something that'd be really useful to backport in 2.x, and do it via parent poms to apply for all Jackson components not just databind.

Reply:
Reading about lifecycle phases, I am bit confused here wrt prepare-package -- doesn't that come in rather late (see https://avajava.com/tutorials/lessons/what-are-the-phases-of-the-maven-default-lifecycle.html), and the originally defined phase of generate-sources should be better choice?

Reply:
Trying to figure out why Idea can not generate sources, found this

https://intellij-support.jetbrains.com/hc/en-us/community/posts/206107279-Maven-support-generated-source-folders-problem

that would suggest problem to be missing directory under `target-generated-sources/`, as supposedly each plug-in would be expected to have intermediate directory there. Not sure that is relevant.
Also noticed that `jackson-base` has settings for Eclipse plug-in that probably tries to solve the same problem, yet I don't think that actually works (importing project without cli build fails for missing `PackageVersion.java`).
Sigh.

Reply:
> Reading about lifecycle phases, I am bit confused here wrt prepare-package

You are right, it looks like `generate-sources` would be the correct choice here. The funny thing is that `maven-replacer-plugin` examples are all using `prepare-package` (maybe because it is intended to generate something else, as its descriptions imply). Still, I'm not sure if Maven is tracking back lifecycle tasks to compile the generated source at a later stage.

The IDEA issue you reported seems ancient at this point, so I guess it was just missing the correct hooks to generate to code at some Maven lifecycle phase. To be honest, using `maven-replacer-plugin` feels old since the plugin does not look maintained anymore.

I did some research and based on the [list of plugins listed here](http://maven.apache.org/plugins/index.html#outside-the-maven-land) and [here](https://www.mojohaus.org/plugins.html), [templating-maven-plugin](https://github.com/mojohaus/templating-maven-plugin) would be a better option.

WDYT?

Reply:
I have no particular attachment to this plugin, and like you said, it does seem to be sort of orphaned.
Unfortunately I did not implement this integration originally so I am not confident I'd be able to do conversions. But if anyone has time and interest, it sounds like a good idea to upgrade to another plug-in.
Use case itself seems straight-forward, and it'd be good to avoid requirement to always do cli build before IDE import.

Reply:
I did some research and apparently [Intellij IDEA is still not able to run generate-sources automatically](https://youtrack.jetbrains.com/issue/IDEA-77583) when importing/reloading maven projects. There is an option to do that manually, though:

![Screen Shot 2020-08-05 at 2 50 14 PM](https://user-images.githubusercontent.com/4576/89452001-02dfac00-d72b-11ea-96c0-0e6db7ee4f90.png)

I guess Eclipse also has a way to handle that, but I'm not sure.

---

Anyway, it looks like it is still worth to move away from `maven-replacer-plugin` in favor of using `templating-maven-plugin`, but I'm not sure how much work is involved in doing so. It looks like one should start by removing [it from oss-parent](https://github.com/FasterXML/oss-parent/blob/e6a38442960b341c474b79c7182cd8994652987d/pom.xml#L206-L216), and add `templating-maven-plugin` there. But that will probably cascade to all subprojects, right?

Reply:
I think that the way to replace use of plugin should probably start by replacing it on specific projects first, for 2.12, and once proven could demote settings to shared parent classes.
At first (2.12) it would be fine to include default settings for the new plug-in of course (version to use, configuration), but require specific Jackson components to enable the new plug-in with explicit inclusion.